src/c++/perf_analyzer/test_concurrency_manager.cc

```diff
@@ -0,0 +1,372 @@
+          doctest::Approx(params.max_concurrency).epsilon(0.25));
+    }
+  }
+
+  void CheckSequences(const PerfAnalyzerParameters& params)
+  {
+    auto stats = GetStats();
+
+    // Make sure all seq IDs are within range
+    //
+    for (auto seq_id : stats->sequence_status.used_seq_ids) {
+      CHECK(seq_id >= params.start_sequence_id);
+      CHECK(seq_id <= params.start_sequence_id + params.sequence_id_range);
+    }
+
+    // Make sure that we had the correct number of concurrently live sequences
+    //
+    // If the sequence length is only 1 then there is nothing to check because
+    // there are never any overlapping requests -- they always immediately exit
+    //
+    if (params.sequence_length != 1) {
+      CHECK(
+          stats->sequence_status.max_live_seq_count == params.max_concurrency);
+    }
+
+    // Make sure that the length of each sequence is as expected
+    // (The code explicitly has a 20% slop, so that is what we are checking)
+    //
+    for (auto len : stats->sequence_status.seq_lengths) {
+      CHECK(len == doctest::Approx(params.sequence_length).epsilon(0.20));
+    }
+  }
+
+  std::unique_ptr<LoadManager> CreateManager(
+      PerfAnalyzerParameters params, bool is_sequence_model = false)
+  {
+    std::unique_ptr<LoadManager> manager;
+    std::shared_ptr<cb::ClientBackendFactory> factory =
+        std::make_shared<cb::MockClientBackendFactory>(stats_);
+    std::shared_ptr<ModelParser> parser =
+        std::make_shared<MockModelParser>(is_sequence_model);
+
+    ConcurrencyManager::Create(
+        params.async, params.streaming, params.batch_size, params.max_threads,
+        params.max_concurrency, params.sequence_length, params.string_length,
+        params.string_data, params.zero_input, params.user_data,
+        params.shared_memory_type, params.output_shm_size,
+        params.start_sequence_id, params.sequence_id_range, parser, factory,
+        &manager);
+
+    return manager;
+  }
+};
+
+/// Test that the correct concurrency is maintained in the load manager
+///
+TEST_CASE("concurrency_infer_type")
+{
+  TestConcurrencyManager tcm{};
+  PerfAnalyzerParameters params{};
+
+  params.max_concurrency = 1;
+
+  SUBCASE("async_streaming")
+  {
+    params.async = true;
+    params.streaming = true;
+  }
+  SUBCASE("async_no_streaming")
+  {
+    params.async = true;
+    params.streaming = false;
+  }
+  SUBCASE("no_async_streaming")
+  {
+    params.async = false;
+    params.streaming = true;
+  }
+  SUBCASE("no_async_no_streaming")
+  {
+    params.async = false;
+    params.streaming = false;
+  }
+
+  tcm.TestInferType(params);
+}
+
+/// Check that the corre
+/// are used given different param values for async and stream
+///
+TEST_CASE("concurrency_concurrency")
+{
+  TestConcurrencyManager tcm{};
+  PerfAnalyzerParameters params{};
+  std::chrono::milliseconds request_delay{50};
+  std::chrono::milliseconds sleep_time{225};
+
+  SUBCASE("")
```

Comment:
name the subcases. It is reported on a failure, which we will need to figure out what instance failed since they all use the same common check code.

Reply:
Added names 👍